docs/subsystems/notifications.md

```diff
@@ -0,0 +1,152 @@
+  * Data about the message's content (E.g. mentions, wildcard
+  mentions, and alert words) and encodes it into the `UserMessage`
+  table's `flags` structure, which is in turn passed into
+  `send_event` for each user receiving the message.
+  * Data about user configuration relevant to the message, such as
+  `push_notify_user_ids` and `stream_notify_user_ids`, are included
+  alongside `flags` in the per-user data structure.
+  * The `presence_idle_user_ids` set, containing the subset of
+  recipient users who are mentioned, are PM recipients, have alert
+  words, or otherwise would normally get a notification, but have not
+  interacted with a Zulip client in the last few minutes.  (Users who
+  have generally will not receive a notification unless the
+  `enable_online_push_notifications` flag is enabled).  This data
+  structure ignores users for whom the message is not notifiable,
+  which is important to avoid this being thousands of `user_ids` for
+  messages to large streams with few currently active users.
+* The Tornado [event queue system](../subsystems/events-system.md)
+  processes that data, as well as data about each user's active event
+  queues, to (1) push an event to each queue needing that message and
+  (2) for notifiable messages, pushing an event onto the
+  `missedmessage_mobile_notifications` and/or `missedmessage_emails`
+  queues. This important message-processing logic has notable extra
+  logic not present when processing normal events, both for details
+  like splicing `flags` to customize event payloads per-user, as well.
+  * The Tornado system determines whether the user is "offline/idle".
+    Zulip's email notifications are designed to not fire when the user
+    is actively using Zulip to avoid spam, and this is where those
+    checks are implemented.
+  * Users in `presence_idle_user_ids` are always considered idle:
+    the variable name means "users who are idle because of
+    presence". This is how we solve the Idle Desktop Problem; users
+    with an idle desktop are treated the same as users who aren't
+    logged in for this check.
+  * However, that check does not handle the Hard Disconnect Problem:
+    if a user was present 1 minute before a message was sent, and then
+    closed their laptop, the user will not be in
+    `presence_idle_user_ids`, and so without an additional mechanism,
+    messages sent shortly after a user leaves would never trigger a
+    notification (!).
+  * We solve that problem by also notifying if
+    `receiver_is_off_zulip`, which checks whether the user has any
+    current events system clients registered to receive `message`
+    events. This check is done immediately (handling soft disconnects,
+    where E.g. the user closes their last Zulip tab and we get the
+    `DELETE /events/{queue_id}` request).
+  * The `receiver_is_off_zulip` check is effectively repeated when
+    event queues are garbage-collected (in `missedmessage_hook`);
```

Comment:
@timabbott AFAIK `receiver_is_off_zulip` is'nt checked when event queues are garbage-collected or maybe I'm missing something?

Reply:
@shanukun I think the key word here is `effectively`. We're not literally calling `receiver_is_off_zulip`, but what happens in `missedmessage_hook` is analogical to it in its role.